### ISSUE TITLE
Copy module takes forever 

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29589]
From @Mihai-CMM on 2016-10-11T13:49:39Z
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

Copy module
##### ANSIBLE VERSION

```
ansible --version
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### OS / ENVIRONMENT

Managing OS: Fedora 22 
Managed node OS: Centos 6.8 VM in Proxmox
##### SUMMARY

Copying one file takes forever. Sometimes ssh connection expires before copy action is finished
##### STEPS TO REPRODUCE

```
- hosts: proxmox
   vars_files:
      - files/configuration/proxcred.yml
      - files/configuration/configuration.yml
    remote_user: root
    tasks:
   - name: copy superadmin.sql function to remote machine
     copy: src="superadmin.sql"  dest="/var/log/superadmin.sql"
```

##### EXPECTED RESULTS

Copy takes a veeeeery long time / ssh connection expires 

Copied from original issue: ansible/ansible-modules-core#5227

**Comment by ansibot** [comment]
From @Mihai-CMM on 2016-10-11T13:49:39Z

date && ansible-playbook -vvvvv populate.yml &&date
Tue Oct 11 16:15:18 EEST 2016
Using /home/mcostache/PROIECTELE_MELE/Proxmox_Ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: populate.yml *********************************************************
1 plays in populate.yml

PLAY [proxmox] *****************************************************************

TASK [setup] *******************************************************************
<10.3.0.60> ESTABLISH SSH CONNECTION FOR USER: root
<10.3.0.60> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<10.3.0.60> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<10.3.0.60> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=root)
<10.3.0.60> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=30)
<10.3.0.60> SSH: PlayContext set ssh_common_args: ()
<10.3.0.60> SSH: PlayContext set ssh_extra_args: ()
<10.3.0.60> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r)
<10.3.0.60> SSH: EXEC sshpass -d12 ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r 10.3.0.60 '/bin/sh -c '"'"'( umask 77 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1476191719.06-107688176639325`" && echo ansible-tmp-1476191719.06-107688176639325="`echo $HOME/.ansible/tmp/ansible-tmp-1476191719.06-107688176639325`" ) && sleep 0'"'"''
<10.3.0.60> PUT /tmp/tmpPAzLmv TO /root/.ansible/tmp/ansible-tmp-1476191719.06-107688176639325/setup
<10.3.0.60> SSH: disable batch mode for sshpass: (-o)(BatchMode=no)
<10.3.0.60> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<10.3.0.60> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<10.3.0.60> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=root)
<10.3.0.60> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=30)
<10.3.0.60> SSH: PlayContext set ssh_common_args: ()
<10.3.0.60> SSH: PlayContext set sftp_extra_args: ()
<10.3.0.60> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r)
<10.3.0.60> SSH: EXEC sshpass -d12 sftp -o BatchMode=no -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r '[10.3.0.60]'
<10.3.0.60> ESTABLISH SSH CONNECTION FOR USER: root
<10.3.0.60> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<10.3.0.60> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<10.3.0.60> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=root)
<10.3.0.60> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=30)
<10.3.0.60> SSH: PlayContext set ssh_common_args: ()
<10.3.0.60> SSH: PlayContext set ssh_extra_args: ()
<10.3.0.60> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r)
<10.3.0.60> SSH: EXEC sshpass -d12 ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r -tt 10.3.0.60 '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1476191719.06-107688176639325/setup; rm -rf "/root/.ansible/tmp/ansible-tmp-1476191719.06-107688176639325/" > /dev/null 2>&1 && sleep 0'"'"''
ok: [10.3.0.60]

TASK [copy superadmin.sql function to remote machine] **************************
task path: /home/mcostache/PROIECTELE_MELE/Proxmox_Ansible/populate.yml:20
<10.3.0.60> ESTABLISH SSH CONNECTION FOR USER: root
<10.3.0.60> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<10.3.0.60> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<10.3.0.60> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=root)
<10.3.0.60> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=30)
<10.3.0.60> SSH: PlayContext set ssh_common_args: ()
<10.3.0.60> SSH: PlayContext set ssh_extra_args: ()
<10.3.0.60> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r)
<10.3.0.60> SSH: EXEC sshpass -d12 ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r 10.3.0.60 '/bin/sh -c '"'"'( umask 77 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1476192632.12-274642950027047`" && echo ansible-tmp-1476192632.12-274642950027047="`echo $HOME/.ansible/tmp/ansible-tmp-1476192632.12-274642950027047`" ) && sleep 0'"'"''
<10.3.0.60> PUT /tmp/tmp616mbR TO /root/.ansible/tmp/ansible-tmp-1476192632.12-274642950027047/stat
<10.3.0.60> SSH: disable batch mode for sshpass: (-o)(BatchMode=no)
<10.3.0.60> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<10.3.0.60> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<10.3.0.60> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=root)
<10.3.0.60> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=30)
<10.3.0.60> SSH: PlayContext set ssh_common_args: ()
<10.3.0.60> SSH: PlayContext set sftp_extra_args: ()
<10.3.0.60> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r)
<10.3.0.60> SSH: EXEC sshpass -d12 sftp -o BatchMode=no -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r '[10.3.0.60]'
fatal: [10.3.0.60]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
    to retry, use: --limit @populate.retry

PLAY RECAP *********************************************************************
10.3.0.60                  : ok=1    changed=0    unreachable=1    failed=0  
Tue Oct 11 16:46:42 EEST 2016

**Comment by ansibot** [comment]
From @ansibot on 2016-10-11T13:49:39Z

@ansible, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @jctanner on 2016-10-11T13:49:39Z

@Mihai-CMM what is the size of the file being copied and how long does it take before the connection times out?

needs_info

**Comment by ansibot** [comment]
From @Mihai-CMM on 2016-10-11T13:49:39Z

@jjctanner   Thank you for looking into this:
 size of the file : 0,5 KB
[mcostache@localhost Proxmox_Ansible]$ ll -h superadmin.sql 
-rwxr-xr-x 1 mcostache mcostache 538 Oct  3 16:02 superadmin.sql

connection times out after  : ~30 minutes
date && ansible-playbook -vvvvv populate.yml &&date

## Tue Oct 11 16:15:18 EEST 2016

10.3.0.60 : ok=1 changed=0 unreachable=1 failed=0
Tue Oct 11 16:46:42 EEST 2016

**Comment by ansibot** [comment]
From @jctanner on 2016-10-11T13:49:39Z

@Mihai-CMM I suspect there is a prompt being displayed that sshpass does not know how to handle or stfp is defunct on the remote machine. Please try testing sftp directly to see if the problem can be reproduced outside of ansible ...

```
sftp -o BatchMode=no -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=root -o ConnectTimeout=30 -o ControlPath=/home/mcostache/.ansible/cp/ansible-ssh-%h-%p-%r '[10.3.0.60]
```

needs_info

**Comment by ansibot** [comment]
From @Mihai-CMM on 2016-10-11T13:49:39Z

@jctanner 
Thanks again. sftp output

[mcostache@localhost Proxmox_Ansible]$ date
Fri Nov 18 14:08:51 EET 2016
[mcostache@localhost Proxmox_Ansible]$ sftp root@10.3.0.60 
root@10.3.0.60's password: 
Connected to 10.3.0.60.
sftp> put superadmin.sql /tmp
Uploading superadmin.sql to /tmp/superadmin.sql
superadmin.sql                                                                                            100%  538     0.5KB/s   00:00  
sftp> exit
[mcostache@localhost Proxmox_Ansible]$ date
Fri Nov 18 14:09:08 EET 2016

**Comment by ansibot** [comment]
From @naggie on 2016-10-11T13:49:39Z

Is a fix planned? I have the same issue. It takes up to an hour or so just to copy a handful of scripts.....

**Comment by ansibot** [comment]
From @rahulx411 on 2016-10-11T13:49:39Z

same here --copying files is the the slowest part of playbook run

**Comment by ansibot** [comment]
From @naggie on 2016-10-11T13:49:39Z

It's a good job the fix is coming in 2.4.0 then :) https://github.com/ansible/ansible/issues/21513#issuecomment-312901102

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (114/100 (113%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext8.iad.github.net ([192.30.252.199]:40296 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ0r-000nLm-Hm
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:48:37 -0500
Date: Mon, 11 Sep 2017 07:47:33 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141268;
	bh=pQeapz11H2ZVeUhG/PTK7UhTW+1WOTW93uoy5maYq2I=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=PpYzj7cKvfOzXRDfPQoJf+xPl3KWQmor7XxTPoy3/7Dc/2/x+RrORgKSHKmLqRmXw
	 6UypRZBNORLFwVMyrs67QZoc4Sr4WbMnPnmawls0BdarltaPqgY4ILWI1fZ8iwoVFm
	 oBEAEJbNQb/CEC81EjQcNNdZegr4gTaRyvfD1L6Q=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (115/100 (114%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext5.iad.github.net ([192.30.252.196]:45817 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ12-000nSa-Pi
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:48:48 -0500
Date: Mon, 11 Sep 2017 07:47:53 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141277;
	bh=KSmCBdvfyeSrYOte8rbTWMQcF5jJ67WAinqnOp7uwqI=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=cXtBtioRm7/4zqNMhfVSlWUTo6UkUeaWbFZFQIaz6aCbegwK8oFn7syuvBVmq5NIZ
	 yyx+5Q+BmwLu+x9bAgqZBixlQEq+EFDz2IxR3THn8Qit0ukDAYlsLC9ETY99BJpe/j
	 EJjMHB4vOuIM6JrWt3uM6vegbvSdnlLlLyYlHvSg=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (116/100 (115%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 12196 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext2.iad.github.net ([192.30.252.193]:46062 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ18-000nVj-EF
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:48:54 -0500
Date: Mon, 11 Sep 2017 07:48:13 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141294;
	bh=Q19vCBYkhlIfKxIV2137TrQEISYVSi0zIGRQrRo//qQ=;
	h=From:Reply-To:To:Cc:Subject:List-ID:List-Archive:List-Post:
	 List-Unsubscribe:From;
	b=UidbVE4FubmSBUVHPUvBPAj0PPuSIgPXxsbsmBgjc25TosXkhbdYlKvK6+m6AXa+2
	 0CesGlDJ311Tz1RQtFmKZb7xkdHBTgMsOyog/XDPM/aKHq+3XAM564m9p+hvzhh6Tb
	 EoGBuTc0Eo1mt42CUZEDHW2wX5UDHT6E7GsWO3xU=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (117/100 (117%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext7.iad.github.net ([192.30.252.198]:49987 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drQ1A-000nX4-OM
	for greg@biomassiv.es; Mon, 11 Sep 2017 09:48:56 -0500
Date: Mon, 11 Sep 2017 07:47:43 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505141283;
	bh=cnQjQJxdKNBuQnZcxRqrydZyXdpeFknFITnhgNXa5iw=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=q2jAfvrkAZBqrdEkYLORqIqNdy/E9jP/HZ4zh3plRRQf331Eobry8S7KRyhuRREvy
	 6e+tfH3KVI1PTABea2KLEg4Eibkqt7C/g+3J99Q7yBNkUFdTVKSNxPfk9G2E/3DXvF
	 v+E4rsPR/vSfqR9X/gcPAjrprFdgF/1xDZ+SxIz0=

**Comment by abadger** [comment]
Is this still a problem?  From looking at it, the original report is not a problem with the copy module per-se but something about the way the connection plugin is copying the file.  I'm not able to reproduce it locally using current ansible, a small file, and sftp as the transport method.

needs_info

**Comment by naggie** [comment]
I think it's an order of magnitude faster than it was, but still slow if you're copying a lot of files such as a git repository or python virtualenv. I suppose this is because of all the checking it does on each file, which may not be pipelined.

I just avoid using it nowadays.

**Comment by abadger** [comment]
<nod> Yeah.  For a large number of files with recursive, there are definitely performance issues (although, as you said, 2.4.0 sped that up considerably).  This bug report was specifically about copying one 538 byte file so it's something else.... Since it doesn't seem to be a reproducible problem in copy, something about the connection seems more likely (for this issue... the recursive issue is definitely a problem in copy's design).

**Comment by Mihai-CMM** [comment]
At that time I've found an workaround. I am not working on that project anymore but i am positive it was a network issue created by DF (don't fragment) flags on the pfsense. A few months later after i've opened this issue we hit other problems with SCP or ssh caused by DF.

**Comment by ansibot** [comment]
cc @ptux
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gsikorski** [comment]
I am not sure if this is manifestation of the same issue, but take a look here: https://sourceforge.net/p/sshpass/patches/11. Also https://github.com/ansible/ansible/issues/14335 seems to be the same issue to me.

**Comment by bcoca** [comment]
Closing this issue as copy itself will never be as fast as direct tools (scp/sftp) and we recommend the syncrhonize module when you are attempting to copy a multitude of files.

The 'hanging' issues are unrelated to slowness so should have a different issue opened.
